src/ComposerJson.php

```diff
@@ -125,6 +144,76 @@ private function extractAutoloadPaths(string $basePath, array $autoload, bool $i
         return $result;
     }
 
+    /**
+     * @param array<string> $exclude
+     * @return array<string, bool>
+     * @throws InvalidPathException
+     */
+    private function extractAutoloadExcludeRegexps(string $basePath, array $exclude, bool $isDev): array
+    {
+        $regexps = [];
+
+        foreach ($exclude as $path) {
+            $regexps[$this->resolveAutoloadExclude($basePath, $path)] = $isDev;
+        }
+
+        return $regexps;
+    }
+
+    /**
+     * Implementation copied from composer/composer.
+     *
+     * @license MIT https://github.com/composer/composer/blob/ee2c9afdc86ef3f06a4bd49b1fea7d1d636afc92/LICENSE
+     * @see https://getcomposer.org/doc/04-schema.md#exclude-files-from-classmaps
+     * @see https://github.com/composer/composer/blob/ee2c9afdc86ef3f06a4bd49b1fea7d1d636afc92/src/Composer/Autoload/AutoloadGenerator.php#L1256-L1286
+     * @throws InvalidPathException
+     */
+    private function resolveAutoloadExclude(string $basePath, string $pathPattern): string
+    {
+        // first escape user input
+        $path = preg_replace('{/+}', '/', preg_quote(trim(strtr($pathPattern, '\\', '/'), '/')));
+
+        if ($path === null) {
+            throw new InvalidPathException("Failure while globbing $pathPattern path.");
+        }
+
+        // add support for wildcards * and **
+        $path = strtr($path, ['\\*\\*' => '.+?', '\\*' => '[^/]+?']);
+
+        // add support for up-level relative paths
+        $updir = null;
+        $path = preg_replace_callback(
+            '{^((?:(?:\\\\\\.){1,2}+/)+)}',
+            static function ($matches) use (&$updir): string {
+                if (isset($matches[1]) && $matches[1] !== '') {
+                    // undo preg_quote for the matched string
+                    $updir = str_replace('\\.', '.', $matches[1]);
+                }
+
+                return '';
+            },
+            $path
+            // note: composer also use `PREG_UNMATCHED_AS_NULL` but it is supported since PHP v7.4
```

Comment:
[It is available since 7.2](https://www.php.net/manual/en/pcre.constants.php). [Proof](https://3v4l.org/9dV1Q)

Reply:
[`preg_replace_callback($flags)` was added in 7.4](https://www.php.net/manual/en/function.preg-replace-callback.php)